pkg/controllers/user/nodecontrolplane/controlplane_endpoint_sync.go

```diff
@@ -0,0 +1,194 @@
+	selectorMap       = map[string][]labels.Selector{
+		"expose-kube-etcd-metrics":      {etcdLabel, masterLabel},
+		"expose-kube-cm-metrics":        {controlPlaneLabel, masterLabel},
+		"expose-kube-scheduler-metrics": {controlPlaneLabel, masterLabel},
+	}
+)
+
+func (controller *ControlPlaneEndpointController) sync(key string, obj *corev1.Node) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil || obj.Labels == nil {
+		return obj, nil
+	}
+
+	endpointMap := make(map[string]struct{})
+	for endpointNames, selectors := range selectorMap {
+		for _, selector := range selectors {
+			if selector.Matches(labels.Set(obj.Labels)) {
+				endpointMap[endpointNames] = struct{}{}
+				break
+			}
+		}
+	}
+
+	endpoints, err := getTargetEndpoints(controller.EndpointLister, endpointMap)
+	if err != nil {
+		return obj, err
+	}
+
+	internalAddress := node.GetNodeInternalAddress(obj)
+
+	for _, endpoint := range endpoints {
+		if endpointHasAddress(endpoint, internalAddress) {
+			continue
+		}
+		newEndpoint := endpoint.DeepCopy()
+		if err := injectAddressToEndpoint(newEndpoint, internalAddress, controller.ServiceLister); err != nil {
+			continue
+		}
+		_, err = controller.Endpoints.Update(newEndpoint)
+		if err != nil {
+			logrus.WithError(err).Warnf("adding node ip for endpoint %s error", endpoint.Name)
+			return obj, err
+		}
+	}
+	return obj, nil
+}
+
+func (controller *ControlPlaneEndpointController) syncEndpoints(key string, obj *corev1.Endpoints) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	selectors, ok := selectorMap[obj.Name]
+	if !ok {
+		return obj, nil
+	}
+
+	nodes, err := getNodes(controller.NodeLister, selectors)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, n := range nodes {
+		controller.NodeController.Enqueue("", n.Name)
+	}
+
+	return obj, nil
+}
+
+func injectAddressToEndpoint(endpoint *corev1.Endpoints, address string, serviceLister v1.ServiceLister) error {
+	subsetIndex := -1
+	var ports []corev1.EndpointPort
+	svc, err := serviceLister.Get("cattle-prometheus", endpoint.Name)
+	if err != nil {
+		logrus.WithError(err).Warn("listing service error")
+		return err
+	}
+	for _, servicePort := range svc.Spec.Ports {
+		ports = append(ports, getEndpointPort(servicePort))
+	}
+	sort.Slice(ports, func(i, j int) bool {
+		return ports[i].Port < ports[j].Port
+	})
+
+	for i, subset := range endpoint.Subsets {
+		var subsetPorts []corev1.EndpointPort
+		copy(subsetPorts, subset.Ports)
+		sort.Slice(subsetPorts, func(i, j int) bool {
+			return subsetPorts[i].Port < subsetPorts[j].Port
+		})
+		if reflect.DeepEqual(subsetPorts, ports) {
+			subsetIndex = i
+			break
+		}
+	}
+
+	if subsetIndex == -1 {
+		endpoint.Subsets = []corev1.EndpointSubset{{
+			Ports: ports,
+		}}
+		subsetIndex = 0
+	}
+
+	endpoint.Subsets[subsetIndex].Addresses = append(endpoint.Subsets[subsetIndex].Addresses, corev1.EndpointAddress{IP: address})
```

Comment:
@orangedeng I just had a question, do we only append new endpoint here and not remove the old one?

Reply:
I refactored it. Always generate the address from nodes so deleted nodes will be removed from the list.